File: Strivers-A2Z-DSA-Sheet/05.Recursion/2.Subsequences Pattern/python/03.Count distinct substrings.py
"""
=============================================================================
  QUESTION: Count Distinct Subsequences (GFG)
=============================================================================

  Find number of distinct subsequences. Answer mod 10^9+7.
  Example: "gfg" → 7 ("","g","f","gf","fg","gg","gfg")

  APPROACH: Pick/Not-Pick + Set for uniqueness.
  Set auto-handles duplicates (e.g. picking 'g' at idx 0 vs idx 2).

  DRY RUN "gf":
    Include 'g', Include 'f' → "gf" | Include 'g', Exclude 'f' → "g"
    Exclude 'g', Include 'f' → "f"  | Exclude 'g', Exclude 'f' → ""
    set = {"gf","g","f",""} → size = 4 ✓

  Time: O(2^N), Space: O(2^N)
=============================================================================
"""

MOD = 10**9 + 7


def solve(s, index, temp, st):
    if index == len(s):
        st.add(temp)
        return

    # Include
    solve(s, index + 1, temp + s[index], st)

    # Exclude
    solve(s, index + 1, temp, st)


def distinct_subsequences(s):
    st = set()
    solve(s, 0, "", st)
    return len(st) % MOD


# Driver Code
if __name__ == "__main__":
    print(distinct_subsequences("gfg"))  # 7
    print(distinct_subsequences("abc"))  # 8

File: Strivers-A2Z-DSA-Sheet/05.Recursion/2.Subsequences Pattern/javascript/03.Count distinct substrings.js
/*
=============================================================================
  QUESTION: Count Distinct Subsequences (GFG)
=============================================================================

  Given string of lowercase alphabets, find number of distinct subsequences.
  Answer modulo 10^9 + 7.

  Examples:
    "gfg" → 7  ("", "g", "f", "gf", "fg", "gg", "gfg")
    "abc" → 8  (all 2^3 = 8 distinct, no duplicates)
    "aa"  → 3  ("", "a", "aa")

=============================================================================
  APPROACH: Pick / Not-Pick + Set — O(2^N) Time, O(2^N) Space
=============================================================================

  Same pick/not-pick as Power Set, but use a Set to eliminate duplicates.

  Why duplicates?  In "gfg": pick idx0('g') skip idx2('g') = "g"
                             skip idx0('g') pick idx2('g') = "g"  ← same!
  Set automatically handles this.

  Dry Run ("gf"):
    solve(0, "") → pick 'g': solve(1, "g") → pick 'f': add "gf"
                                             → skip 'f': add "g"
                 → skip 'g': solve(1, "")  → pick 'f': add "f"
                                             → skip 'f': add ""
    Set = {"gf", "g", "f", ""} → size = 4 ✓

=============================================================================
*/

function distinctSubsequences(s) {
    let set = new Set();

    function solve(index, temp) {
        if (index === s.length) {
            set.add(temp);
            return;
        }
        solve(index + 1, temp + s[index]);  // Include
        solve(index + 1, temp);              // Exclude
    }

    solve(0, "");
    return set.size % (1e9 + 7);
}

// ==========================================================================
// DRIVER CODE
// ==========================================================================
console.log(distinctSubsequences("gfg"));  // 7
console.log(distinctSubsequences("abc"));  // 8
console.log(distinctSubsequences("aa"));   // 3
